ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,570 @@
+func (rs *traceRulesSampler) limit() (float64, bool) {
+	if rs.enabled() {
+		return float64(rs.limiter.limiter.Limit()), true
+	}
+	return math.NaN(), false
+}
+
+// defaultRateLimit specifies the default trace rate limit used when DD_TRACE_RATE_LIMIT is not set.
+const defaultRateLimit = 100.0
+
+// newRateLimiter returns a rate limiter which restricts the number of traces sampled per second.
+// The limit is DD_TRACE_RATE_LIMIT if set, `defaultRateLimit` otherwise.
+func newRateLimiter() *rateLimiter {
+	limit := defaultRateLimit
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v != "" {
+		l, err := strconv.ParseFloat(v, 64)
+		if err != nil {
+			log.Warn("DD_TRACE_RATE_LIMIT invalid, using default value %f: %v", limit, err)
+		} else if l < 0.0 {
+			log.Warn("DD_TRACE_RATE_LIMIT negative, using default value %f", limit)
+		} else {
+			// override the default limit
+			limit = l
+		}
+	}
+	return &rateLimiter{
+		limiter:  rate.NewLimiter(rate.Limit(limit), int(math.Ceil(limit))),
+		prevTime: time.Now(),
+	}
+}
+
+// singleSpanRulesSampler allows a user-defined list of rules to apply to spans
+// to sample single spans.
+// These rules match based on the span's Service and Name. If empty value is supplied
+// to either Service or Name field, it will default to "*", allow all.
+// When making a sampling decision, the rules are checked in order until
+// a match is found.
+// If a match is found, the rate from that rule is used.
+// If no match is found, no changes or further sampling is applied to the spans.
+// The rate is used to determine if the span should be sampled, but an upper
+// limit can be defined using the max_per_second field when supplying the rule.
+// If max_per_second is absent in the rule, the default is allow all.
+// Its value is the max number of spans to sample per second.
+// Spans that matched the rules but exceeded the rate limit are not sampled.
+type singleSpanRulesSampler struct {
+	rules []SamplingRule // the rules to match spans with
+}
+
+// newSingleSpanRulesSampler configures a *singleSpanRulesSampler instance using the given set of rules.
+// Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
+func newSingleSpanRulesSampler(rules []SamplingRule) *singleSpanRulesSampler {
+	return &singleSpanRulesSampler{
+		rules: rules,
+	}
+}
+
+func (rs *singleSpanRulesSampler) enabled() bool {
+	return len(rs.rules) > 0
+}
+
+// apply uses the sampling rules to determine the sampling rate for the
+// provided span. If the rules don't match, then it returns false and the span is not
+// modified.
+func (rs *singleSpanRulesSampler) apply(span *span) bool {
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			rs.applyRate(span, rule, rule.Rate, time.Now())
+			return true
+		}
+	}
+	return false
+}
+
+func (rs *singleSpanRulesSampler) applyRate(span *span, rule SamplingRule, rate float64, now time.Time) {
```

Comment:
Why do you need this separate method? Why not put all of this code inside `apply`?

Reply:
Not sure if you forgot this comment. If you feel strongly on having this additional method, it's ok, but I don't see the reason for it. If you keep it, maybe rename it like this:
```suggestion
func (rs *singleSpanRulesSampler) applyRule(span *span, rule SamplingRule, now time.Time) {
```
There is no point to have the `rate` argument since it's derived from `rule.Rate`